### PR TITLE
AIInterface, Threat handling and Movement Rewrite

### PR DESCRIPTION
**Pull request by aaron02**
**Description**
Rewrite all AI Handling to get rid of all Hackfixes, add clearness to it and make it better than now. Full movement rewrite to use newSpline System and Movement Generators from TrinityCore. Added Propper Threat handling. After this PR Transport npcs work as intendet and dont glitch around.

fixes:
#380
#634
#224 
#516 
#606
#708

If possible take your Time and test some stuff and report any bugs you find :)

**Todo / Checklist**
- [x] classic build without scripts.
- [x] tbc build without scripts.
- [x] wotlk build without scripts.
- [x] cata build without scripts.
- [x] mop build without scripts.
- [x] fix scripts for all build targets.
- [x] fix pathfinding (atm creatures fly around when enabled).
- [x] creatures should not move while rooted
- [x] creatures should not attack while polymorphed if you are near them
- [x] creatures no longer fly even when they have flight set to 1 in creature_properties_movement (tested with bat guards in Thrallmar)
- [x] flying creatures should fall down when dying
- [x] flying creatures no longer fall down when respawned and killed again
- [x] flying creatures should not fall through buildings (https://github.com/AscEmu/AscEmu/issues/224)
- [x] creature won't always evade when you run very far away from it (.char set speed 40, melee attack mob once, run very far like 200 or 300 yards away) sometimes it evades but mostly it won't
- [x] player cannot move after entering on transport (dismounting on transport may cause this)
- [x] knockback effects are inverted
- [x] remove all UseNewAIInterface defines
- [x] if sheeped/polymorphed, creature will turn his facing to you when confused movement is paused
- [x] creatures won't evade if they cannot reach you
- [x] swimming creatures won't follow you and keep evading (to test .worldport 0 -11880.261719 -374.873901 -10.124239 and fight against the fishes there)
- [x] some flying creatures won't even aggro you (to test .worldport 571 7414.893066 2107.205566 505.081421 and fight against the flying gargoyles there)
- [x] some flying creatures won't fall down on death (same mobs as above)
- [x] while underwater, creatures won't update z position
- [x] movement in tbc does not work

### CODE REVIEW

**Comment by Appled** [comment]
Possibly fixing issues https://github.com/AscEmu/AscEmu/issues/634 https://github.com/AscEmu/AscEmu/issues/606 https://github.com/AscEmu/AscEmu/issues/516

**Comment by Zyres** [comment]
(WIP) Review functionality:
- [ ] Command .waypoint add does not work (Tested on Cata)

**Comment by aaron02** [comment]
Npc Posses has nothing to do with this PR

**Comment by aaron02** [comment]
https://user-images.githubusercontent.com/1266630/120069451-56d6cb00-c086-11eb-99fe-b4c8a3a6e0e4.mp4

cata follow and npc come is working for me lol

**Comment by aaron02** [comment]
i belive it must be some sort of databse issue with the looping creatures:
image

the grunt left is looping the othe rone stopped at my position

File: src/world/Movement/WaypointManager.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+{
+    auto oldMSTime = Util::TimeNow();
+    _waypointStore.clear();
+
+    //                                                0    1         2           3          4            5           6        7      8           9
+    QueryResult* result = WorldDatabase.Query("SELECT id, point, position_x, position_y, position_z, orientation, move_type, delay, action, action_chance FROM creature_waypoints ORDER BY id, point");
+
+    if (!result)
+    {
+        sLogger.info("WaypointMgr >> Loaded 0 waypoints. DB table `waypoint_data` is empty!");
```
[the remark this answers is not in the document]
Comment (aaron02) [reply src/world/Movement/WaypointManager.cpp]: i changed the table name and then i changed it back and forgot there

**Comment by schnek** [comment]
>  (Maybe unrelated) Thunderbluff transporter npcs on the ground. Maybe duplicate spawns and unrelated :-) (Tested on Cata)

all versions

>  After relog on a transport you will fall to the ground instead on the transport. (tested on Cata)

+wotlk

**Comment by aaron02** [comment]
image
however i dont manage to get these spawns on the ground :(

**Comment by schnek** [comment]
[3874af9]
image
image
npc from spawn transport disappear after a while (or when the cycle ends)

and freezing of transport

https://user-images.githubusercontent.com/1216225/120076460-433f5a80-c0ae-11eb-9541-58e4b7466bea.mp4

**Comment by aaron02** [comment]
transports are mostly client driven. so for example you saw a transport it gets loaded. now if u are a bit to far away it unloads on the server and it should send a update packet so the client destroys it visualy. now here is the case that u wont receive that packet maybe out of mapmgr update distance idk. and the transport visualy stays on ur client until you relog. try with my settings i set my active cell count to 5 cells default is 1

**Comment by schnek** [comment]
ok ok but the video is also about spawn 😁😀 and where are the settings in the client or in the server?

**Comment by aaron02** [comment]
in the world.conf MapCellNumber        = "5"

**Comment by aaron02** [comment]
transport handling needs a bit more work to make it better, buildupdatevalues should send the pathprogress todisplay them correctly for all clients for example this is a bigger problem when using them in instances or battlegrounds

**Comment by aaron02** [comment]
and the transport u postet has double spawns

File: src/world/Movement/WaypointManager.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+void WaypointMgr::deleteAllWayPoints(uint32_t pathid)
+{
+    WorldDatabase.Execute("DELETE FROM creature_waypoints WHERE id = %u", pathid);
+
+    auto itr = _waypointStore.find(pathid);
+    if (itr != _waypointStore.end())
+        _waypointStore.erase(pathid);
+
+    sLogger.debug("Deleted waypoints for pathID %u", pathid);
+}
```
Comment (Appled) [inline src/world/Movement/WaypointManager.cpp:164]: Missing newline

File: src/world/Objects/Object.h [hunk 3]
```diff
@@ -462,6 +506,32 @@ class SERVER_DECL Object : public EventableObject, public IUpdatable
...
+        float getRelativeAngle(LocationVector const& pos) const { return toRelativeAngle(getAbsoluteAngle(pos.x, pos.y)); }
+        float getRelativeAngle(LocationVector const* pos) const { return toRelativeAngle(getAbsoluteAngle(pos)); }
+
+        bool isInDist(Object* pos, float dist) { return GetPosition().getExactDistSq(pos->GetPositionX(), pos->GetPositionY(), pos->GetPositionZ()) < dist * dist; }
+
+        void getNearPoint2D(Object* searcher, float& x, float& y, float distance, float absAngle);
+        void getNearPoint(Object* searcher, float& x, float& y, float& z, float distance2d, float absAngle);
+        void getClosePoint(float& x, float& y, float& z, float size, float distance2d = 0, float relAngle = 0);
+
+        LocationVector GetHitSpherePointFor(LocationVector const& dest);
```
Comment (Appled) [inline src/world/Objects/Object.h:524]: Functions should start with lowercase character

File: src/world/Objects/Object.h [hunk 4]
```diff
@@ -462,6 +506,32 @@ class SERVER_DECL Object : public EventableObject, public IUpdatable
...
+        void GetHitSpherePointFor(LocationVector const& dest, float& x, float& y, float& z) const;
```
Comment (Appled) [inline src/world/Objects/Object.h:525]: Functions should start with lowercase character

File: src/world/Objects/Object.h [hunk 5]
```diff
@@ -462,6 +506,32 @@ class SERVER_DECL Object : public EventableObject, public IUpdatable
...
+        LocationVector GetHitSpherePointFor(LocationVector const& dest) const;
```
Comment (Appled) [inline src/world/Objects/Object.h:526]: Functions should start with lowercase character

File: src/world/Objects/Object.h [hunk 6]
```diff
@@ -462,6 +506,32 @@ class SERVER_DECL Object : public EventableObject, public IUpdatable
...
+        void UpdateAllowedPositionZ(float x, float y, float &z, float* groundZ = nullptr);
```
Comment (Appled) [inline src/world/Objects/Object.h:527]: Functions should start with lowercase character

File: src/world/Objects/Object.h [hunk 7]
```diff
@@ -462,6 +506,32 @@ class SERVER_DECL Object : public EventableObject, public IUpdatable
...
+        float GetMapWaterOrGroundLevel(float x, float y, float z, float* ground = nullptr);
```
Comment (Appled) [inline src/world/Objects/Object.h:528]: Functions should start with lowercase character

File: src/world/Objects/Object.h [hunk 8]
```diff
@@ -462,6 +506,32 @@ class SERVER_DECL Object : public EventableObject, public IUpdatable
...
+        void MovePositionToFirstCollision(LocationVector &pos, float dist, float angle);
```
Comment (Appled) [inline src/world/Objects/Object.h:529]: Functions should start with lowercase character

File: src/world/Objects/ObjectMgr.cpp [hunk 9]
```diff
@@ -3306,3 +3196,68 @@ void ObjectMgr::LoadInstanceEncounters()
...
+    if (creatureMovement.Chase >= CreatureChaseMovementType::Max)
+    {
+        creatureMovement.Chase = CreatureChaseMovementType::Run;
+    }
+
+    if (creatureMovement.Random >= CreatureRandomMovementType::Max)
+    {
+        creatureMovement.Random = CreatureRandomMovementType::Walk;
+    }
+}
```
Comment (Appled) [inline src/world/Objects/ObjectMgr.cpp:3263]: Missing newline

File: src/world/Units/Creatures/AIInterface.h [hunk 10]
```diff
@@ -107,538 +88,334 @@ enum AI_SpellTargetType
...
-
-        Movement::WaypointMovementScript mWaypointScriptType;
-        int32_t mNextPoint;
-        bool mWaitTimerSetOnWP;
+    // AI Agent functions
+    bool m_canRangedAttack;
+    void selectCurrentAgent(Unit* target, uint32_t spellid);
+    void initializeSpells();
+
+    void SetCannotReachTarget(bool cannotReach);
```
Comment (Appled) [inline src/world/Units/Creatures/AIInterface.h:145]: Functions should start with lowercase character

File: src/world/Units/Creatures/AIInterface.h [hunk 11]
```diff
@@ -107,538 +88,334 @@ enum AI_SpellTargetType
...
+    bool CanNotReachTarget() const { return m_cannotReachTarget; }
```
Comment (Appled) [inline src/world/Units/Creatures/AIInterface.h:146]: Functions should start with lowercase character

File: src/world/Units/Creatures/AIInterface.h [hunk 12]
```diff
@@ -107,538 +88,334 @@ enum AI_SpellTargetType
...
-        bool mMoveWaypointsBackwards;
+private:
+    uint32_t m_AiState;
 
-        bool mShowWayPoints;
-        bool mShowWayPointsBackwards;
+    //////////////////////////////////////////////////////////////////////////////////////////
+    // Combat functions
+public:
+    void JustEnteredCombat(Unit* pUnit);
```
Comment (Appled) [inline src/world/Units/Creatures/AIInterface.h:204]: Functions should start with lowercase character

File: src/world/Units/Creatures/AIInterface.h [hunk 13]
```diff
@@ -107,538 +88,334 @@ enum AI_SpellTargetType
...
+
+    bool canUnitEvade(unsigned long time_passed);
+    void enterEvadeMode();
+    bool _enterEvadeMode();
+
+    void initGroupThreat(Unit* target);
+    void instanceCombatProgress(bool activate);
+
+    // Event Handler
+    void HandleEvent(uint32_t event, Unit* pUnit, uint32_t misc1);
```
Comment (Appled) [inline src/world/Units/Creatures/AIInterface.h:294]: Functions should start with lowercase character

File: src/world/Units/Creatures/AIInterface.h [hunk 14]
```diff
@@ -107,538 +88,334 @@ enum AI_SpellTargetType
...
+    void EventUnwander(Unit* pUnit, uint32_t misc1);
+    void EventWander(Unit* pUnit, uint32_t misc1);
+    void EventUnfear(Unit* pUnit, uint32_t misc1);
+    void EventFear(Unit* pUnit, uint32_t misc1);
+    void EventFollowOwner(Unit* pUnit, uint32_t misc1);
+    void EventDamageTaken(Unit* pUnit, uint32_t misc1);
+    void EventLeaveCombat(Unit* pUnit, uint32_t misc1);
+    void EventEnterCombat(Unit* pUnit, uint32_t misc1);
+    void EventChangeFaction(Unit* ForceAttackersToHateThisInstead = NULL);    /// we have to tell our current enemies to stop attacking us, we should also forget about our targets
+    void OnDeath(Object* pKiller);
```
Comment (Appled) [inline src/world/Units/Creatures/AIInterface.h:308]: Functions should start with lowercase character (includes all event functions above)

File: src/world/Units/Creatures/AIInterface.h [hunk 15]
```diff
@@ -107,538 +88,334 @@ enum AI_SpellTargetType
...
+
+    // Update
+    void Update(unsigned long time_passed);
```
Comment (Appled) [inline src/world/Units/Creatures/AIInterface.h:311]: Functions should start with lowercase character

File: src/world/Units/Creatures/AIInterface.h [hunk 16]
```diff
@@ -107,538 +88,334 @@ enum AI_SpellTargetType
...
-
-        void UpdateSpeeds();
-
-        void UpdateMovementSpline();
-
-        void OnMoveCompleted();
-
-        void MoveEvadeReturn();
+    // Pet functions
+    inline void SetPetOwner(Unit* owner) { m_PetOwner = owner; }
```
Comment (Appled) [inline src/world/Units/Creatures/AIInterface.h:371]: Functions should start with lowercase character

File: src/world/Units/Creatures/Creature.h [hunk 17]
```diff
@@ -272,18 +291,37 @@ class SERVER_DECL Creature : public Unit
...
-        void FormationLinkUp(uint32 SqlId);
         void ChannelLinkUpGO(uint32 SqlId);
         void ChannelLinkUpCreature(uint32 SqlId);
-        bool haslinkupevent;
-        Movement::WayPoint* CreateWaypointStruct();
+
         uint32 spawnid;
         uint32 original_emotestate;
 
+        void Motion_Initialize();
```
Comment (Appled) [inline src/world/Units/Creatures/Creature.h:313]: Functions should start with lowercase character

File: src/world/Units/Creatures/Creature.h [hunk 18]
```diff
@@ -272,18 +291,37 @@ class SERVER_DECL Creature : public Unit
...
+        void immediateMovementFlagsUpdate();
+        void updateMovementFlags();
+        CreatureMovementData const& GetMovementTemplate();
+        bool CanWalk() { return GetMovementTemplate().IsGroundAllowed(); }
```
Comment (Appled) [inline src/world/Units/Creatures/Creature.h:317]: Functions should start with lowercase character

File: src/world/Units/Creatures/Creature.h [hunk 19]
```diff
@@ -272,18 +291,37 @@ class SERVER_DECL Creature : public Unit
...
+        bool CanSwim() override { return GetMovementTemplate().IsSwimAllowed() || isPet(); }
```
Comment (Appled) [inline src/world/Units/Creatures/Creature.h:318]: Functions should start with lowercase character

File: src/world/Units/Creatures/Creature.h [hunk 20]
```diff
@@ -272,18 +291,37 @@ class SERVER_DECL Creature : public Unit
...
+        bool CanFly()  override { return GetMovementTemplate().IsFlightAllowed() || IsFlying(); }
```
Comment (Appled) [inline src/world/Units/Creatures/Creature.h:319]: Functions should start with lowercase character

File: src/world/Units/Creatures/Creature.h [hunk 21]
```diff
@@ -272,18 +291,37 @@ class SERVER_DECL Creature : public Unit
...
+        bool CanHover() { return GetMovementTemplate().Ground == CreatureGroundMovementType::Hover || IsHovering(); }
```
Comment (Appled) [inline src/world/Units/Creatures/Creature.h:320]: Functions should start with lowercase character

File: src/world/Units/Creatures/Creature.h [hunk 22]
```diff
@@ -272,18 +291,37 @@ class SERVER_DECL Creature : public Unit
...
+
+        MovementGeneratorType GetDefaultMovementType() const override { return m_defaultMovementType; }
```
Comment (Appled) [inline src/world/Units/Creatures/Creature.h:322]: Functions should start with lowercase character

File: src/world/Units/Creatures/Creature.h [hunk 23]
```diff
@@ -272,18 +291,37 @@ class SERVER_DECL Creature : public Unit
...
+        void SetDefaultMovementType(MovementGeneratorType mgt) { m_defaultMovementType = mgt; }
```
Comment (Appled) [inline src/world/Units/Creatures/Creature.h:323]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureDefines.hpp [hunk 24]
```diff
@@ -41,6 +41,63 @@ const uint32 creatureMaxInventoryItems = 150;
...
+        Random(CreatureRandomMovementType::Walk) { }
+
+    CreatureGroundMovementType Ground;
+    CreatureFlightMovementType Flight;
+    bool Swim;
+    bool Rooted;
+    CreatureChaseMovementType Chase;
+    CreatureRandomMovementType Random;
+
+    bool IsGroundAllowed() const { return Ground != CreatureGroundMovementType::None; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureDefines.hpp:92]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureDefines.hpp [hunk 25]
```diff
@@ -41,6 +41,63 @@ const uint32 creatureMaxInventoryItems = 150;
...
+    bool IsSwimAllowed() const { return Swim; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureDefines.hpp:93]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureDefines.hpp [hunk 26]
```diff
@@ -41,6 +41,63 @@ const uint32 creatureMaxInventoryItems = 150;
...
+    bool IsFlightAllowed() const { return Flight != CreatureFlightMovementType::None; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureDefines.hpp:94]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureDefines.hpp [hunk 27]
```diff
@@ -41,6 +41,63 @@ const uint32 creatureMaxInventoryItems = 150;
...
+    bool IsRooted() const { return Rooted; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureDefines.hpp:95]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureDefines.hpp [hunk 28]
```diff
@@ -41,6 +41,63 @@ const uint32 creatureMaxInventoryItems = 150;
...
+
+    CreatureChaseMovementType GetChase() const { return Chase; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureDefines.hpp:97]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureDefines.hpp [hunk 29]
```diff
@@ -41,6 +41,63 @@ const uint32 creatureMaxInventoryItems = 150;
...
+    CreatureRandomMovementType GetRandom() const { return Random; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureDefines.hpp:98]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureGroups.h [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+        uint32_t _leaderSpawnId;
+        bool _formed;
+        bool _engaging;
+
+    public:
+        //Group cannot be created empty
+        explicit CreatureGroup(uint32_t leaderSpawnId);
+        ~CreatureGroup();
+
+        Creature* GetLeader() const { return _leader; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureGroups.h:69]: Functions should start with lowercase character

File: src/world/Units/Creatures/CreatureGroups.h [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+        uint32_t GetLeaderSpawnId() const { return _leaderSpawnId; }
```
Comment (Appled) [inline src/world/Units/Creatures/CreatureGroups.h:70]: Functions should start with lowercase character

File: src/world/Units/ThreatHandler.cpp [hunk 32]
```diff
@@ -0,0 +1,823 @@
...
+    auto it = _threatenedByMe.find(guid);
+    if (it != _threatenedByMe.end())
+        _threatenedByMe.erase(it);
+}
+
+void ThreatManager::heapNotifyChanged()
+{
+    if (_sortedThreatList.size())
+        _sortedThreatList.sort([](ThreatReference* a, ThreatReference* b) -> bool { return a->getThreat() > b->getThreat(); });
+}
```
Comment (Appled) [inline src/world/Units/ThreatHandler.cpp:823]: Missing newline

File: src/world/Units/ThreatHandler.h [hunk 33]
```diff
@@ -0,0 +1,212 @@
...
+    int32 _tempModifier;
+    TauntState _taunted;
+
+public:
+    ThreatReference(ThreatReference const&) = delete;
+    ThreatReference& operator=(ThreatReference const&) = delete;
+
+    friend class ThreatManager;
+    friend struct CompareThreatLessThan;
+};
```
Comment (Appled) [inline src/world/Units/ThreatHandler.h:212]: Missing newline

File: src/world/Units/Unit.cpp [hunk 34]
```diff
@@ -5551,3 +5974,285 @@ uint64_t Unit::getTransGuid()
...
+
+MovementGeneratorType Unit::GetDefaultMovementType() const
+{
+    return IDLE_MOTION_TYPE;
+}
+
+bool Unit::isSplineEnabled() const
+{
+    return movespline->Initialized();
+}
```
Comment (Appled) [inline src/world/Units/Unit.cpp:6258]: Missing newline

File: src/world/Units/Unit.h [hunk 35]
```diff
@@ -586,7 +590,29 @@ class SERVER_DECL Unit : public Object
     bool isTurning() const { return obj_movement_info.hasMovementFlag(MOVEFLAG_TURNING_MASK); }
     bool IsFlying() const { return obj_movement_info.hasMovementFlag(MOVEFLAG_FLYING_MASK); }
     bool IsFalling() const;
-    virtual bool CanSwim() const;
+    virtual bool CanSwim();
```
Comment (Appled) [inline src/world/Units/Unit.h:593]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 36]
```diff
@@ -586,7 +590,29 @@ class SERVER_DECL Unit : public Object
...
+    bool isUnderWater() const;
+    bool isInAccessiblePlaceFor(Creature* c) const;
+
+    uint64_t getCharmerOrOwnerGUID() const override { return isCharmed() ? getCharmedByGuid() : getSummonedByGuid(); }
+    bool isCharmed() const { return !getCharmedByGuid(); }
+
+    void setControlled(bool apply, UnitStates state);
+    void applyControlStatesIfNeeded();
+
+    virtual bool CanFly();
```
Comment (Appled) [inline src/world/Units/Unit.h:604]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 37]
```diff
@@ -586,7 +590,29 @@ class SERVER_DECL Unit : public Object
...
+
+    bool IsWalking() const { return obj_movement_info.hasMovementFlag(MOVEFLAG_WALK); }
```
Comment (Appled) [inline src/world/Units/Unit.h:606]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 38]
```diff
@@ -586,7 +590,29 @@ class SERVER_DECL Unit : public Object
...
+    bool IsHovering() const { return obj_movement_info.hasMovementFlag(MOVEFLAG_HOVER); }
```
Comment (Appled) [inline src/world/Units/Unit.h:607]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 39]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
 
+    void propagateSpeedChange();
+    void setSpeedRate(UnitSpeedType mtype, float rate, bool current);
+
+    uint8_t m_forced_speed_changes[MAX_SPEED_TYPE];
+
     // Movement info
     MovementNew::MoveSpline* movespline;
 
+    void FollowerAdded(AbstractFollower* f) { m_followingMe.insert(f); }
```
Comment (Appled) [inline src/world/Units/Unit.h:636]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 40]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+    void FollowerRemoved(AbstractFollower* f) { m_followingMe.erase(f); }
```
Comment (Appled) [inline src/world/Units/Unit.h:637]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 41]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+    void RemoveAllFollowers();
```
Comment (Appled) [inline src/world/Units/Unit.h:638]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 42]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+    virtual float getFollowAngle() const { return static_cast<float>(M_PI / 2); }
+
+    MovementManager* getMovementManager() { return i_movementManager; }
+    MovementManager const* getMovementManager() const { return i_movementManager; }
+
+    void StopMoving();
```
Comment (Appled) [inline src/world/Units/Unit.h:644]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 43]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+    void PauseMovement(uint32_t timer = 0, uint8_t slot = 0, bool forced = true); // timer in ms
```
Comment (Appled) [inline src/world/Units/Unit.h:645]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 44]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+    void ResumeMovement(uint32_t timer = 0, uint8_t slot = 0); // timer in ms
```
Comment (Appled) [inline src/world/Units/Unit.h:646]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 45]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+
+private:
+    std::unordered_set<AbstractFollower*> m_followingMe;
+
+protected:
+    MovementManager* i_movementManager;
+
+    void SetFeared(bool apply);
```
Comment (Appled) [inline src/world/Units/Unit.h:654]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 46]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+    void SetConfused(bool apply);
```
Comment (Appled) [inline src/world/Units/Unit.h:655]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 47]
```diff
@@ -596,13 +622,39 @@ class SERVER_DECL Unit : public Object
...
+    void SetStunned(bool apply);
```
Comment (Appled) [inline src/world/Units/Unit.h:656]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 48]
```diff
@@ -613,6 +665,9 @@ class SERVER_DECL Unit : public Object
 public:
     void sendMoveSplinePaket(UnitSpeedType speed_type);
     void disableSpline();
+    bool isSplineEnabled() const;
+
+    virtual MovementGeneratorType GetDefaultMovementType() const;
```
Comment (Appled) [inline src/world/Units/Unit.h:670]: Functions should start with lowercase character

File: src/world/Units/Unit.h [hunk 49]
```diff
@@ -1376,6 +1434,7 @@ class SERVER_DECL Unit : public Object
     virtual void Die(Unit* pAttacker, uint32 damage, uint32 spellid);
     virtual bool isCritter() { return false; }
 
+    void KnockbackFrom(float x, float y, float speedXY, float speedZ);
```
Comment (Appled) [inline src/world/Units/Unit.h:1437]: Functions should start with lowercase character

**Review by Appled — CHANGES_REQUESTED** [review]

File: src/world/Units/Creatures/AIInterface.h (hunk 15 above)
Comment (aaron02) [reply src/world/Units/Creatures/AIInterface.h:311]: this is inerhitted from iUpdatable

File: src/world/Map/MapMgr.h [hunk 50]
```diff
@@ -172,6 +175,10 @@ class SERVER_DECL MapMgr : public CellHandler <MapCell>, public EventableObject,
 
     uint8 GetLiquidType(float x, float y);
 
+    bool isUnderWater(float x, float y, float z);
+
+    ZLiquidStatus GetLiquidStatus(uint32 phaseMask, float x, float y, float z, uint8 ReqLiquidType, LiquidData* data = nullptr, float collisionHeight = 2.03128f); // DEFAULT_COLLISION_HEIGHT in Object.h
```
Comment (Appled) [inline src/world/Map/MapMgr.h:180]: You forgot one :P

**Review by Appled — APPROVED** [review]
All good, just fix that one function 👍

**Pull request merged**